Library/Dependencies/napi/source/js_native_api_JavaScriptCore.cc

```diff
@@ -1,97 +1,220 @@
+    size_t errorStrSize{ 0 };
+
+    if (!JSValueIsNull(env->m_globalContext, exception))
+    {
+        // Getting error string
+        if (JSValueIsString(env->m_globalContext, exception))
+        {
+            OpaqueJSValue* ncstr = const_cast<OpaqueJSValue*>(exception);
+            /*size_t length =*/ JSStringGetUTF8CString(reinterpret_cast<JSStringRef>(ncstr), errorStr, sizeof(errorStr));
+        }
+        else
+        {
+            OpaqueJSValue *exceptionValue = const_cast<OpaqueJSValue*>(exception);
+            napi_get_value_string_utf8(env, reinterpret_cast<napi_value>(exceptionValue), errorStr, sizeof(errorStr), &errorStrSize);
+        }
+    }
+}
 
-napi_status napi_close_escapable_handle_scope(napi_env env,
-  napi_escapable_handle_scope scope) {
-    // stub
+napi_status napi_get_value_bool(napi_env env, napi_value v, bool* result) {
+    bool boolean = JSValueToBoolean(env->m_globalContext, reinterpret_cast<JSValueRef>(v));
+    *result = boolean;
+    return napi_ok;
+}
+
+napi_status napi_get_value_double(napi_env env, napi_value v, double* result) {
+    double dblValue = JSValueToNumber(env->m_globalContext, reinterpret_cast<JSValueRef>(v), NULL);
+    *result = dblValue;
+    return napi_ok;
+}
+
+napi_status napi_get_value_int32(napi_env env, napi_value v, int32_t* result) {
+    double dblValue = JSValueToNumber(env->m_globalContext, reinterpret_cast<JSValueRef>(v), NULL);
+    *result = int32_t(dblValue);
+    return napi_ok;
+}
+
+napi_status napi_get_value_uint32(napi_env env,
+                                  napi_value v,
+                                  uint32_t* result) {
+    double dblValue = JSValueToNumber(env->m_globalContext, reinterpret_cast<JSValueRef>(v), NULL);
+    *result = uint32_t(dblValue);
     return napi_ok;
 }
 
 napi_status napi_create_double(napi_env env,
                                double value,
                                napi_value* result) {
-    // stub
+    OpaqueJSValue* jsValue = const_cast<OpaqueJSValue*>(JSValueMakeNumber(env->m_globalContext, value));
+    *result = reinterpret_cast<napi_value>(jsValue);
     return napi_ok;
 }
 
 napi_status napi_create_error(napi_env env,
                               napi_value code,
                               napi_value msg,
                               napi_value* result) {
-    // stub
+    JSObjectRef errorObject = JSObjectMakeError(env->m_globalContext, 0, NULL, NULL);
+    *result = reinterpret_cast<napi_value>(errorObject);
     return napi_ok;
 }
 
 napi_status napi_create_reference(napi_env env,
                                   napi_value v,
                                   uint32_t initial_refcount,
                                   napi_ref* result) {
-    // stub
+    auto value = reinterpret_cast<JSValueRef>(v);
+    auto info = new RefInfo
+    {
+        value,
+        initial_refcount
+    };
+    
+    if (info->count != 0)
+    {
+        //CHECK_JSRT(JsAddRef(value, nullptr)); todo
```

Comment:
we need something to track that this implementation isn't complete. maybe add something to the roadmap or create an issue?

Reply:
I created issues for JSC and metal. 
https://github.com/BabylonJS/BabylonNative/issues/54
https://github.com/BabylonJS/BabylonNative/issues/53